Explaining controllers as viable alternative (#13554)
symfony/symfony-docs: security/voters.rst

How to Use Voters to Check User Permissions
===========================================

Voters are Symfony's most powerful way of managing permissions. They allow you
to centralize all permission logic, then reuse them in many places.

However, if you don't reuse permissions or your rules are basic, you can always
put that logic directly into your controller instead. Here's an example how
this could look like, if you want to make a route accessible to the "owner" only::

    // src/AppBundle/Controller/PostController.php
    // ...

    if ($post->getOwner() !== $this->getUser()) {
        throw $this->createAccessDeniedException();
    }

> In that sense, the following example used throughout this page is more like a

Reviewer: I am fine with adding the explanation above, but I do not agree with this sentence. The example used below IS a real-world use case. I suggest to remove this sentence.

Author: I removed the second part - please take a look again :-)

For more information take a look at
:ref:`the section about access decision managers <components-security-access-decision-manager>`.

.. tip::

    The example used throughout this page features just two routes (``post_show`` and ``post_edit``).
    However, the advantage of voters is that you can reuse them in *many* places and centralize

>     all permission logic. If you don't reuse permissions and the rules are simple, you instead

Reviewer 2: ```suggestion
    all permission logic. If you don't reuse permissions or the rules are basic, you instead
```

    might want to do the check in the controller directly and throw an ``AccessDeniedException``

>     to create the correct response::

Reviewer 3: I'd say we move this to the VERY top of this chapter - like the very first words. Then we can say something simpler:

```rst
If you need to deny access in a controller based on custom rules - perhaps only the "owner" of an object can access a page - you can always put that logic directly in your controller::

    ... the code

But if you need to repeat this logic in multiple places, a better option is to centralize that logic with a custom voter.
```

And then it would go into the current beginning "In order to use voters".

I think you're right to talk about this - so why not discuss the easy option (and when it might not work) immediately? :)

Author: I tried my best :-)

        // src/AppBundle/Controller/PostController.php
        use Symfony\Component\Security\Core\Exception\AccessDeniedException;

        // ...
        if ($post->getOwner() !== $this->getUser()) {

>             throw new AccessDeniedException();

Reviewer 3: Let's use: `throw $this-> createAccessDeniedException()` - we always assume users are using the normal shortcuts - like `AbstractController`.

Author: Done :-)
